src/gui/transferlistfilterswidget.cpp

```diff
@@ -220,70 +213,99 @@ StatusFilterWidget::StatusFilterWidget(QWidget *parent, TransferListWidget *tran
     const Preferences *const pref = Preferences::instance();
     setCurrentRow(pref->getTransSelFilter(), QItemSelectionModel::SelectCurrent);
     toggleFilter(pref->getStatusFilterState());
+
+    populate();
+
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::torrentsUpdated
+            , this, &StatusFilterWidget::handleTorrentsUpdated);
 }
 
 StatusFilterWidget::~StatusFilterWidget()
 {
     Preferences::instance()->setTransSelFilter(currentRow());
 }
 
-void StatusFilterWidget::updateTorrentNumbers()
+void StatusFilterWidget::populate()
 {
-    int nbDownloading = 0;
-    int nbSeeding = 0;
-    int nbCompleted = 0;
-    int nbResumed = 0;
-    int nbPaused = 0;
-    int nbActive = 0;
-    int nbInactive = 0;
-    int nbStalled = 0;
-    int nbStalledUploading = 0;
-    int nbStalledDownloading = 0;
-    int nbChecking = 0;
-    int nbErrored = 0;
+    m_torrentsStatus.clear();
 
     const QVector<BitTorrent::Torrent *> torrents = BitTorrent::Session::instance()->torrents();
     for (const BitTorrent::Torrent *torrent : torrents)
     {
-        if (torrent->isDownloading())
-            ++nbDownloading;
-        if (torrent->isUploading())
-            ++nbSeeding;
-        if (torrent->isCompleted())
-            ++nbCompleted;
-        if (torrent->isResumed())
-            ++nbResumed;
-        if (torrent->isPaused())
-            ++nbPaused;
-        if (torrent->isActive())
-            ++nbActive;
-        if (torrent->isInactive())
-            ++nbInactive;
-        if (torrent->state() ==  BitTorrent::TorrentState::StalledUploading)
-            ++nbStalledUploading;
-        if (torrent->state() ==  BitTorrent::TorrentState::StalledDownloading)
-            ++nbStalledDownloading;
-        if (torrent->isChecking())
-            ++nbChecking;
-        if (torrent->isErrored())
-            ++nbErrored;
+        updateTorrentStatus(torrent);
     }
 
-    nbStalled = nbStalledUploading + nbStalledDownloading;
+    updateTexts();
+}
+
+void StatusFilterWidget::updateTorrentStatus(const BitTorrent::Torrent *torrent)
+{
+    const auto update = [this, torrent](const int status, const bool insert, int &count)
+    {
+        const bool contains = m_torrentsStatus.contains(torrent, status);
```

Comment:
I think this has repetitive look up on `m_torrentsStatus` for the same torrent. Can it be further optimized?

Reply:
> I think this has repetitive look up on `m_torrentsStatus` for the same torrent. Can it be further optimized?

IMO, QMultiHash isn't good decision here.
I would declare some struct with boolean fields for each status and store it in QSet.

Reply:
This is down from 12% to .2%, I think it's fast enough now, no need to complicate code.

![image](https://user-images.githubusercontent.com/34717789/156939524-c6d1a984-438c-4a64-b485-974fb5c85ae9.png)

![image](https://user-images.githubusercontent.com/34717789/156939576-634bc493-f313-4ba3-b37d-995fd36747d5.png)

Reply:
At least I would  change `QMultiHash<const BitTorrent::Torrent *, int>` with `QHash<const BitTorrent::Torrent *, QSet<int>>`.  This shouldn't complicate it much, but it will get rid of "repetitive look up".

@Chocobo1, it can be improved by someone else later if there are no blocking issues with it.

Reply:
>@Chocobo1, it can be improved by someone else later if there are no blocking issues with it.

I don't mind it deferred. However, I also don't understand why defer it when it is easy to implement it and has advantages.

Reply:
said concerns are imaginary and has negative impact on performance (using QSet)

Reply:
>At least I would change QMultiHash<const BitTorrent::Torrent *, int> with QHash<const BitTorrent::Torrent *, QSet<int>>.

I think a bitfield instead of `QSet<int>` would be even better.

Reply:
>I think a bitfield instead of QSet<int> would be even better.

Why don't you take this opportunity yourself and chase  <0.1% performance difference

Reply:
> However, I also don't understand why defer it when it is easy to implement and has advantages.

Well, at least it should be faster if the idea is embodied by the one to whom it came.